File: java/src/main/java/io/darkstrata/credentialcheck/exception/RateLimitException.java
package io.darkstrata.credentialcheck.exception;

/**
 * Exception thrown when rate limited by the API (429).
 */
public class RateLimitException extends DarkStrataException {

    private final Integer retryAfter;

    public RateLimitException() {
        this(null);
    }

    public RateLimitException(Integer retryAfter) {
        super(
            retryAfter != null
                ? "Rate limited. Retry after " + retryAfter + " seconds"
                : "Rate limited",
            ErrorCode.RATE_LIMIT_ERROR,
            429,
            true
        );
        this.retryAfter = retryAfter;
    }

    /**
     * Get the number of seconds to wait before retrying, if provided by the API.
     */
    public Integer getRetryAfter() {
        return retryAfter;
    }
}

File: credential-check/csharp/src/DarkStrata.CredentialCheck/Exceptions/RateLimitException.cs
namespace DarkStrata.CredentialCheck;

/// <summary>
/// Exception thrown when rate limit is exceeded.
/// </summary>
public sealed class RateLimitException : DarkStrataException
{
    /// <summary>
    /// Seconds until rate limit resets (if available).
    /// </summary>
    public int? RetryAfter { get; }

    /// <summary>
    /// Creates a new RateLimitException.
    /// </summary>
    /// <param name="retryAfter">Seconds until rate limit resets (if available).</param>
    public RateLimitException(int? retryAfter = null)
        : base(
            retryAfter.HasValue
                ? $"Rate limit exceeded. Retry after {retryAfter} seconds."
                : "Rate limit exceeded.",
            ErrorCode.RateLimitError,
            statusCode: 429,
            isRetryable: true)
    {
        RetryAfter = retryAfter;
    }
}
